src/BiDirectionalLinkedListImp.h

```diff
@@ -0,0 +1,152 @@
+//
+// Created by shvmo on 02/05/2020.
+//
+
+#ifndef DATASTRUCTURESWET1_BIDIRECTIONALLINKEDLISTIMP_H
+#define DATASTRUCTURESWET1_BIDIRECTIONALLINKEDLISTIMP_H
+
+#include <cassert>
+#include "BiDirectionalLinkedList.h"
+
+using std::make_shared;
+
+template<class Type>
+BiDirectionalLinkedList<Type>::BiDirectionalLinkedList() = default;
+
+template<class Type>
+BiDirectionalLinkedList<Type>::~BiDirectionalLinkedList() = default;
+
+template<class Type>
+shared_ptr<BiDirectionalNode<Type>> BiDirectionalLinkedList<Type>::getFirst() const {
+    return this->first;
+}
+
+template<class Type>
+BiDirectionalNode<Type> *BiDirectionalLinkedList<Type>::getFirstPointer() const {
+    return this->first.get();
+}
+
+template<class Type>
+shared_ptr<BiDirectionalNode<Type>> BiDirectionalLinkedList<Type>::getLast() const {
+    return this->last;
+}
+
+template<class Type>
+void BiDirectionalLinkedList<Type>::insert(const Type &newValue) {
+    if (this->first == nullptr) {
+        this->first = make_shared<BiDirectionalNode<Type>>();
+
+        this->first->setNext(nullptr);
+        this->first->setPrevious(nullptr);
+        this->first->setValue(newValue);
+
+        this->last = this->first;
+    } else {
+        shared_ptr<BiDirectionalNode<Type>> newLast = make_shared<BiDirectionalNode<Type>>();
+
+        newLast->setNext(nullptr);
+        newLast->setPrevious(last);
+        newLast->setValue(newValue);
+
+        last->setNext(newLast);
+
+        this->last = newLast;
+    }
+}
```

Comment:
i think this version is less duplication:
![image](https://user-images.githubusercontent.com/26043145/81996479-d188a000-9655-11ea-9c56-e0b5a5c9603e.png)

what do you think?

Reply:
couldn't you copy paste the code? why print screen :(

Reply:
`void BiDirectionalLinkedList<Type>::insert(const Type &newValue) {
    shared_ptr<BiDirectionalNode<Type>> newLast = make_shared<BiDirectionalNode<Type>>();
    
    newLast->setValue(newValue);
    newLast->setNext(nullptr);
    
    if (this->first == nullptr) {
        newLast->setPrevious(nullptr);
        
        this->first = newLast;
        this->last = this->first;
    } else {
        newLast->setPrevious(this->last);
        this->last->setNext(newLast);

        this->last = newLast;
    }
}`

Reply:
fixxed